packages/client/src/internals/dataLoader.ts

```diff
@@ -26,100 +37,125 @@ type BatchLoader<TKey, TValue> = {
     dispatchTimer = null;
-    batch = null;
+    pendingItems = null;
   };
-  function dispatch() {
-    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
-    const thisBatch = batch!;
-    destroyTimerAndBatch();
 
-    const itemBatches: BatchItem<TKey, TValue>[][] = [[]];
+  /**
+   * Iterate through the items and split them into groups based on the `batchLoader`'s validate function
+   */
+  function groupItems(items: BatchItem<TKey, TValue>[]) {
+    const groupedItems: BatchItem<TKey, TValue>[][] = [[]];
+    let index = 0;
+    while (true) {
+      const item = items[index];
+      if (!item) {
+        // we're done
+        break;
+      }
+      const lastGroup = groupedItems[groupedItems.length - 1];
+
+      if (item.aborted) {
+        // Item was aborted before it was dispatched
+        item.reject(new Error('Aborted'));
+        index++;
+        continue;
+      }
 
-    for (const item of thisBatch.items) {
-      const lastItemBatch = itemBatches[itemBatches.length - 1];
       const isValid = batchLoader.validate(
-        lastItemBatch.concat(item).map((_item) => _item.key),
+        lastGroup.concat(item).map((it) => it.key),
       );
+
       if (isValid) {
-        lastItemBatch.push(item);
-      } else {
-        itemBatches.push([item]);
-        if (lastItemBatch.length === 0) {
-          itemBatches.push([]);
-        }
+        lastGroup.push(item);
+        index++;
+        continue;
+      }
+
+      if (lastGroup.length === 0) {
+        item.reject(new Error('Input is too big for a single dispatch'));
+        index++;
+        continue;
       }
+      // Create new group, next iteration will try to add the item to that
+      groupedItems.push([]);
     }
+    return groupedItems;
+  }
+
+  function dispatch() {
+    const groupedItems = groupItems(pendingItems!);
+    destroyTimerAndPendingItems();
 
-    const promises: Promise<void>[] = [];
-    const cancels: CancelFn[] = [];
+    // Create batches for each group of items
+    for (const items of groupedItems) {
+      if (!items.length) {
+        continue;
+      }
+      const batch: Batch<TKey, TValue> = {
+        items,
+        cancel: throwFatalError,
+      };
+      for (const item of items) {
+        item.batch = batch;
+      }
+      const { promise, cancel } = batchLoader.fetch(
+        batch.items.map((_item) => _item.key),
+      );
+      batch.cancel = cancel;
 
-    for (const itemBatch of itemBatches) {
-      if (itemBatch.length > 0) {
-        const { promise, cancel } = batchLoader.fetch(
-          itemBatch.map((_item) => _item.key),
-        );
-        const thenPromise = promise.then((result) => {
+      promise
+        .then((result) => {
           for (let i = 0; i < result.length; i++) {
             const value = result[i];
-            itemBatch[i].resolve(value);
+            const item = batch.items[i];
+            item.resolve(value);
+            item.batch = null;
```

Comment:
This is just to "unlock" it in case this circular dependency between the item and the batch would cause some memory leaks. Might be unnecessary, but doesn't hurt